File: internal/order/order.go
package order

import (
	"bufio"
	"cmp"
	"context"
	"encoding/csv"
	"fmt"
	"io"
	"iter"
	"net/url"
	"runtime/trace"
	"strconv"

	"golang.org/x/sync/errgroup"
)

type Order struct {
	Duration     int     `json:"duration"`
	IsBuyOrder   bool    `json:"is_buy_order"`
	Issued       string  `json:"issued"`
	LocationID   int     `json:"location_id"`
	MinVolume    int     `json:"min_volume"`
	OrderID      int     `json:"order_id"`
	Price        float64 `json:"price"`
	Range        string  `json:"range"`
	SystemID     int     `json:"system_id"`
	TypeID       int     `json:"type_id"`
	VolumeRemain int     `json:"volume_remain"`
	VolumeTotal  int     `json:"volume_total"`
}

func (o Order) Record() [12]string {
	return [12]string{strconv.Itoa(o.Duration),
		strconv.FormatBool(o.IsBuyOrder),
		o.Issued,
		strconv.Itoa(o.LocationID),
		strconv.Itoa(o.MinVolume),
		strconv.Itoa(o.OrderID),
		strconv.FormatFloat(o.Price, 'f', -1, 64),
		o.Range,
		strconv.Itoa(o.SystemID),
		strconv.Itoa(o.TypeID),
		strconv.Itoa(o.VolumeRemain),
		strconv.Itoa(o.VolumeTotal),
	}
}

type Client interface {
	Regions(ctx context.Context, u *url.URL) iter.Seq2[uint64, error]
	Max(ctx context.Context, u *url.URL, region uint64) (uint64, error)
	Orders(ctx context.Context, u *url.URL, region, page uint64) iter.Seq2[Order, error]
}

const defaultLimit = 1
const defaultBufSize = 0 //1

type Handler struct {
	Client Client
}

func (h *Handler) OrderStream(ctx context.Context, u *url.URL, hasHeader bool) io.ReadCloser {
	g, ctx := errgroup.WithContext(ctx)

	regions := make(chan uint64, defaultBufSize)
	g.Go(func() error {
		ctx, task := trace.NewTask(ctx, "regions")
		defer func() { close(regions); task.End() }()

		for id, err := range h.Client.Regions(ctx, u) {
			if err != nil {
				return err
			}
			wait := trace.StartRegion(ctx, "wait")
			select {
			case <-ctx.Done():
				wait.End()
				return ctx.Err()
			case regions <- id:
				wait.End()
			}
		}
		return nil
	})

	type query struct{ region, page uint64 }
	queries := make(chan query, defaultBufSize)
	g.Go(func() error {
		defer close(queries)

		g, ctx := errgroup.WithContext(ctx)
		g.SetLimit(defaultLimit)
		for region := range regions {
			g.Go(func() error {
				ctx, task := trace.NewTask(ctx, "pages")
				defer task.End()

				n, err := h.Client.Max(ctx, u, region)
				if err != nil {
					return fmt.Errorf("failed to fetch max page: %v", err)
				} // else if n < 1

				for i := range n {
					wait := trace.StartRegion(ctx, "wait")
					select {
					case <-ctx.Done():
						wait.End()
						return ctx.Err()
					case queries <- query{region, i + 1}:
						wait.End()
					}
				}
				return nil
			})
		}
		return g.Wait()
	})

	records := make(chan [12]string, defaultBufSize)
	g.Go(func() error {
		defer close(records)

		// send the header if hasHeader is set
		if hasHeader {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case records <- [12]string{
				"duration",
				"is_buy_order",
				"issued",
				"location_id",
				"min_volume",
				"order_id",
				"price",
				"range",
				"system_id",
				"type_id",
				"volume_remain",
				"volume_total",
			}:
			}
		}

		g, ctx := errgroup.WithContext(ctx)
		g.SetLimit(defaultLimit)
		for q := range queries {
			g.Go(func() error {
				ctx, task := trace.NewTask(ctx, "orders")
				defer task.End()

				for o, err := range h.Client.Orders(ctx, u, q.region, q.page) {
					if err != nil {
						return err
					}
					wait := trace.StartRegion(ctx, "wait")
					select {
					case <-ctx.Done():
						wait.End()
						return ctx.Err()
					case records <- o.Record():
						wait.End()
					}
				}
				return nil
			})
		}
		return g.Wait()
	})

	pr, pw := io.Pipe()
	g.Go(func() error {
		ctx, task := trace.NewTask(ctx, "csv")
		defer task.End()

		bw := bufio.NewWriterSize(pw, 32<<10)
		cw := csv.NewWriter(bw) // 4*1<<10 buffer
		for record := range records {
			if err := cmp.Or(cw.Write(record[:]), ctx.Err()); err != nil {
				return err
			}
		}
		cw.Flush()
		return cw.Error()
	})
	go func() { pw.CloseWithError(g.Wait()) }()
	return pr
}

File: src/order.rs
use async_stream::try_stream;
use async_trait::async_trait;
use bytes::Bytes;
use csv_async::AsyncWriterBuilder;
use futures_util::{Stream, StreamExt as _, TryStreamExt};
use serde::{Deserialize, Serialize};
use std::fmt::Debug;
use tokio::{io::duplex, sync::mpsc, task::JoinSet};
use tokio_stream::wrappers::ReceiverStream;
use tokio_util::io::ReaderStream;
use tracing::{Instrument, trace_span};
use url::Url;

const DEFAULT_LIMIT: usize = 1;
const DEFAULT_BUF_SIZE: usize = 1;
const DEFAULT_BUF: usize = 4 << 10;

#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct Order {
    duration: i64,
    is_buy_order: bool,
    issued: String,
    location_id: i64,
    min_volume: i64,
    order_id: i64,
    price: f64,
    range: String,
    system_id: i64,
    type_id: i64,
    volume_remain: i64,
    volume_total: i64,
}

impl Order {
    fn to_record(&self) -> [String; 12] {
        [
            self.duration.to_string(),
            self.is_buy_order.to_string(),
            self.issued.clone(),
            self.location_id.to_string(),
            self.min_volume.to_string(),
            self.order_id.to_string(),
            self.price.to_string(),
            self.range.clone(),
            self.system_id.to_string(),
            self.type_id.to_string(),
            self.volume_remain.to_string(),
            self.volume_total.to_string(),
        ]
    }
}

#[async_trait]
pub trait Client: Clone + Send + Sync + 'static {
    async fn regions(&self, url: &Url) -> Result<impl Stream<Item = Result<u32>> + Send + Unpin>;
    async fn max_pages(&self, url: &Url, region: u32) -> Result<u32>;
    async fn orders(
        &self,
        url: &Url,
        region: u32,
        page: u32,
    ) -> Result<impl Stream<Item = Result<Order>> + Send + Unpin>;
}

#[derive(Debug, Clone)]
pub struct Handler<C> {
    client: C,
}

impl<C> Handler<C>
where
    C: Client,
{
    pub fn new(client: C) -> Self {
        Self { client }
    }

    pub fn order_stream(
        &self,
        base_url: Url,
        has_header: bool,
    ) -> impl Stream<Item = Result<Bytes>> + 'static {
        let mut set = JoinSet::new();

        let (tx, regions) = mpsc::channel(DEFAULT_BUF_SIZE);
        set.spawn({
            let client = self.client.clone();
            let base_url = base_url.clone(); // we form the string here?
            async move {
                let mut stream = client.regions(&base_url).await?;
                while let Some(id) = stream.try_next().await? {
                    tx.send(id).await?;
                }
                Ok(())
            }
            .instrument(trace_span!("regions"))
        });

        let (tx, queries) = mpsc::channel(DEFAULT_BUF_SIZE);
        set.spawn({
            let client = self.client.clone();
            let base_url = base_url.clone();
            async move {
                ReceiverStream::new(regions)
                    .map(Ok::<_, anyhow::Error>)
                    .try_for_each_concurrent(DEFAULT_LIMIT, async |region| {
                        async {
                            let last = client.max_pages(&base_url, region).await?;
                            for page in 1..=last {
                                tx.send((region, page)).await?;
                            }
                            Ok(())
                        }
                        .instrument(trace_span!("pages", region))
                        .await
                    })
                    .await
            }
        });

        let (tx, mut records) = mpsc::channel(DEFAULT_BUF_SIZE);
        set.spawn({
            let client = self.client.clone();
            let base_url = base_url.clone();
            async move {
                if has_header {
                    tx.send([
                        "duration".to_string(),
                        "is_buy_order".to_string(),
                        "issued".to_string(),
                        "location_id".to_string(),
                        "min_volume".to_string(),
                        "order_id".to_string(),
                        "price".to_string(),
                        "range".to_string(),
                        "system_id".to_string(),
                        "type_id".to_string(),
                        "volume_remain".to_string(),
                        "volume_total".to_string(),
                    ])
                    .await?
                }
                ReceiverStream::new(queries)
                    .map(Ok)
                    .try_for_each_concurrent(DEFAULT_LIMIT, async |(region, page)| {
                        async {
                            let mut stream = client.orders(&base_url, region, page).await?;
                            while let Some(order) = stream.try_next().await? {
                                tx.send(order.to_record()).await?;
                            }
                            Ok(())
                        }
                        .instrument(trace_span!("orders", region, page))
                        .await
                    })
                    .await
            }
        });

        let (rx, tx) = duplex(DEFAULT_BUF);
        set.spawn({
            async move {
                let mut wri = AsyncWriterBuilder::new()
                    .buffer_capacity(DEFAULT_BUF)
                    .create_writer(tx);
                while let Some(order) = records.recv().await {
                    wri.write_record(&order).await?;
                }
                wri.flush().await?;
                Ok(())
            }
            .instrument(trace_span!("csv"))
        });

        //  AsyncStream<Result<Bytes, Error>, impl Future<Output = ()>>
        let stream = try_stream! {
            let mut stream = ReaderStream::new(rx);
            while let Some(msg) = stream.next().await {
                yield msg?
            }
            while let Some(res) = set.join_next().await {
                res??;
            }
        };
        stream
    }
}

// DefaultClient

pub type Result<T, E = anyhow::Error> = core::result::Result<T, E>;

#[allow(non_snake_case)]
#[inline]
pub fn Ok<T, E>(value: T) -> Result<T, E> {
    Result::Ok(value)
}
